### PR TITLE
Registers `User` entity as dependency of `getMe`

### PR DESCRIPTION
**Pull request by infomiho**
Fixes https://github.com/wasp-lang/wasp/issues/1110

We didn't register it previously and it led to a bug above.

It works now, but please review it @sodic @shayneczyzewski 

https://user-images.githubusercontent.com/2223680/230638670-028eb132-55cb-406a-addb-69451fa673ab.mov

### CODE REVIEW

File: waspc/data/Generator/templates/react-app/src/queries/core.d.ts [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
+import { Route } from '../types';
 import type { Expand, _Awaited } from '../universal/types'
 
 export function createQuery<BackendQuery extends GenericBackendQuery>(
   queryRoute: string,
   entitiesUsed: any[]
 ): QueryFor<BackendQuery>
 
+export function addMetadataToQuery(
+  query: (...args: any[]) => Promise<unknown>,
```
Comment (infomiho) [inline waspc/data/Generator/templates/react-app/src/queries/core.d.ts:11]: @sodic The only thing I'm unsure about, please check if I've typed it as "any async function"

File: waspc/data/Generator/templates/react-app/src/auth/useAuth.ts [hunk 2]
```diff
@@ -23,5 +26,8 @@ export async function getMe(): Promise<User | null> {
   }
 }
 
-getMe.queryCacheKey = ['auth/me']
-getMe.route = { method: HttpMethod.Get, path: '/auth/me' }
+addMetadataToQuery(getMe, {
+  relativeQueryPath: getMePath,
```
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/react-app/src/auth/useAuth.ts:30]: Shouldn't the `relativeQueryPath` not begin with the `/` (for use in the `queryCacheKey`?

File: waspc/data/Generator/templates/react-app/src/test/vitest/helpers.tsx [hunk 3]
```diff
@@ -8,9 +8,7 @@ import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
 import { beforeAll, afterEach, afterAll } from 'vitest'
 import { Query } from '../../queries'
 import config from '../../config'
-import { HttpMethod } from '../../types'
-
-export type Route = { method: HttpMethod; path: string }
```
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/react-app/src/test/vitest/helpers.tsx:13]: Should we re-export this from here for safety? Unsure if I used it in docs or anywhere else.

File: waspc/src/Wasp/Generator/WebAppGenerator/AuthG.hs [hunk 4]
```diff
@@ -54,8 +55,14 @@ genCreateAuthRequiredPage auth =
...
-genUseAuth :: Generator FileDraft
-genUseAuth = return $ C.mkTmplFd (C.asTmplFile [relfile|src/auth/useAuth.ts|])
+genUseAuth :: AS.Auth.Auth -> Generator FileDraft
+genUseAuth auth = return $ C.mkTmplFdWithData [relfile|src/auth/useAuth.ts|] tmplData
+  where
+    tmplData = object ["getMeDependsOnEntities" .= ("[" ++ userEntitiyStr ++ "]")]
+    userEntitiyStr = "'" ++ userEntityName ++ "'"
+    userEntityName = AS.refName $ AS.Auth.userEntity auth
+
+-- genUseAuth = return $ C.mkTmplFd (C.asTmplFile [relfile|src/auth/useAuth.ts|])
```
Comment (shayneczyzewski) [inline waspc/src/Wasp/Generator/WebAppGenerator/AuthG.hs:65]: Can nix this comment probably.

**Review by shayneczyzewski — COMMENTED** [review]
Nice job so far @infomiho! First pass through code. I will do some testing.

File: waspc/data/Generator/templates/react-app/src/auth/useAuth.ts (hunk 2 above)
Comment (shayneczyzewski) [reply waspc/data/Generator/templates/react-app/src/auth/useAuth.ts:30]: I mean I guess it's not a strict requirement, but for consistency with ops?

File: waspc/data/Generator/templates/react-app/src/test/vitest/helpers.tsx (hunk 3 above)
Comment (infomiho) [reply waspc/data/Generator/templates/react-app/src/test/vitest/helpers.tsx:13]: Reexported 👍

File: waspc/data/Generator/templates/react-app/src/auth/useAuth.ts (hunk 2 above)
Comment (infomiho) [reply waspc/data/Generator/templates/react-app/src/auth/useAuth.ts]: Change it 👍

File: waspc/data/Generator/templates/react-app/src/auth/useAuth.ts [hunk 5]
```diff
@@ -23,5 +27,8 @@ export async function getMe(): Promise<User | null> {
   }
 }
 
-getMe.queryCacheKey = ['auth/me']
-getMe.route = { method: HttpMethod.Get, path: '/auth/me' }
+addMetadataToQuery(getMe, {
```
Comment (sodic) [inline waspc/data/Generator/templates/react-app/src/auth/useAuth.ts:30]: Placing this on the module's top level does work, but feels a bit janky. Maybe we should put the entire initialization process in a function that returns `getMe`, and then export that.

File: waspc/src/Wasp/Generator/WebAppGenerator/AuthG.hs [hunk 6]
```diff
@@ -54,8 +55,12 @@ genCreateAuthRequiredPage auth =
 -- | Generates React hook that Wasp developer can use in a component to get
 --   access to the currently logged in user (and check whether user is logged in
 --   ot not).
-genUseAuth :: Generator FileDraft
-genUseAuth = return $ C.mkTmplFd (C.asTmplFile [relfile|src/auth/useAuth.ts|])
+genUseAuth :: AS.Auth.Auth -> Generator FileDraft
+genUseAuth auth = return $ C.mkTmplFdWithData [relfile|src/auth/useAuth.ts|] tmplData
+  where
+    tmplData = object ["entitiesGetMeDependsOn" .= ("[" ++ userEntitiyStr ++ "]")]
+    userEntitiyStr = "'" ++ userEntityName ++ "'"
```
Comment (sodic) [inline waspc/src/Wasp/Generator/WebAppGenerator/AuthG.hs:62]: Small typo in the variable name.

**Review by sodic — CHANGES_REQUESTED** [review]
One typo and a few possible small changes.

File: waspc/data/Generator/templates/react-app/src/queries/core.d.ts (hunk 1 above)
Comment (infomiho) [reply waspc/data/Generator/templates/react-app/src/queries/core.d.ts:11]: This function receives _some function_ that it attaches additional props to. It was usually defined inside of `makeQuery` and it isn't really connected with `Query`

**Review by sodic — APPROVED** [review]

**Pull request merged**